Add keyboard shortcuts for moving blocks

ntsekouras opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/master/docs/contributors/repository-management.md#pull-requests. -->

## Description
<!-- Please describe what you have changed or added -->
Resolves #10455

This PR adds keyboard shortcuts for moving blocks up and down.

Keyboard combinations for macOS:
- Move up: <kbd>⌥</kbd><kbd>⌘</kbd><kbd>⇧</kbd><kbd>T</kbd>
- Move down: <kbd>⌥</kbd><kbd>⌘</kbd><kbd>⇧</kbd><kbd>Y</kbd>

Windows/Linux:
- Move up: <kbd>Ctrl</kbd>+<kbd>Alt</kbd>+<kbd>Shift</kbd>+<kbd>T</kbd>
- Move down: <kbd>Ctrl</kbd>+<kbd>Alt</kbd>+<kbd>Shift</kbd>+<kbd>Y</kbd>

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

## Screenshots <!-- if applicable -->

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

ntsekouras committed:
start shortcuts

ntsekouras committed:
Add e2e tests

ntsekouras committed:
fix dependency

youknowriad commented:
This is a really cool PR. I agree that we need to find better key combinations. cc @mcsf maybe.

I got this on safari because of the current shortcuts

<img width="555" alt="Capture d’écran 2020-06-22 à 2 28 44 PM" src="https://user-images.githubusercontent.com/272444/85293111-c9155780-b494-11ea-8fb8-3b216332e2d3.png">

ntsekouras commented:
Thank you @youknowriad! I will change them for now to something that works for all ( Mac, Linux, Windows ), just for testing purposes.

I will search first and then make a proposal for key combination, but everyone is welcome to share their thoughts.

@mcsf

ntsekouras committed:
temporary change of key combination for test purposes

gziolo commented on packages/e2e-test-utils/src/multi-select-blocks.js:
```diff
@@ -0,0 +1,18 @@
+ * @param {string} start Identifier of the start block.
+ * @param {string} end Identifier of the end block.
+ */
+export async function multiSelectBlocks( start, end ) {
```
Good idea 👍

Do you think it could be further abstracted, at the moment it’s only a proxy call to data layer and there is a helper for that already:


https://github.com/WordPress/gutenberg/blob/3156d0bbec4983bab009f7fe5dd1b5282df37d18/packages/e2e-test-utils/src/wp-data-select.js#L10

ntsekouras replied:
There is still the need for having a `starting` reference. For example select from `selectMulti( 2, 5 )` that would select from the second block. I'll look into it! Nice idea @gziolo. I should also improve it by selecting the `start` block here.

ntsekouras committed:
Add wpDataDispatch e2e util. Fix wpDataSelect args error.

Mark functions that use wpDataSelect as async as they return Promise. It should be clear how to consume them.

ntsekouras committed:
Add multiSelectBlocksByRange e2e util

mcsf commented:
> I will change them for now to something that works for all ( Mac, Linux, Windows ), just for testing purposes.
> 
> I will search first and then make a proposal for key combination, but everyone is welcome to share their thoughts.

Gutenberg already has shortcuts for "Insert a new block **before** the selected block(s)." (<kbd><kbd>⌥</kbd>+<kbd>⌘</kbd>+<kbd>T</kbd></kbd>) and "Insert a new block **after** the selected block(s)." (<kbd><kbd>⌥</kbd>+<kbd>⌘</kbd>+<kbd>Y</kbd></kbd>). Could we reuse this pattern of <kbd>T</kbd> representing "above" and <kbd>Y</kbd> "below"? Maybe be adding a <kbd>Shift</kbd> to the shortcuts?

gziolo commented on packages/e2e-test-utils/README.md:
```diff
@@ -360,6 +360,33 @@ _Returns_
+
+_Usage_
+
+// selects three blocks -- from the second block to fourth (2, 3, 4).
```
Something is wrong with the example in JSDoc.

ntsekouras replied:
Thanks! Fixed it.

ntsekouras committed:
change example jsdoc

ntsekouras committed:
Change keyboard combination

ntsekouras commented:
>Could we reuse this pattern of T representing "above" and Y "below"? Maybe be adding a Shift to the shortcuts?

@mcsf I believe this is great idea, as it would be intuitive to the user. I implemented this for test drive :)

mcsf commented on packages/block-editor/src/components/keyboard-shortcuts/index.js:
```diff
@@ -12,15 +12,24 @@ import { __ } from '@wordpress/i18n';
+				getBlockRootClientId,
+			} = select( 'core/block-editor' );
+			const selectedClientIds = getSelectedBlockClientIds();
+			const normalizedClientIds = castArray( selectedClientIds );
```
Looking at [`getSelectedBlockClientIds`](https://github.com/WordPress/gutenberg/blob/77171699b76f8ebc3b0d552f0d84fc2ec650a231/packages/block-editor/src/store/selectors.js#L635-L683), it always returns an array. Any reason to array-cast into a new variable?

ntsekouras replied:
You're right! Thanks! I removed it.

mcsf commented on packages/block-editor/src/components/keyboard-shortcuts/index.js:
```diff
@@ -220,6 +257,26 @@ function KeyboardShortcutsRegister() {
+		registerShortcut( {
+			name: 'core/block-editor/move-up',
+			category: 'block',
+			description: __( 'Move block Up.' ),
```
```suggestion
			description: __( 'Move the selected block(s) up.' ),
```

mcsf commented on packages/block-editor/src/components/keyboard-shortcuts/index.js:
```diff
@@ -220,6 +257,26 @@ function KeyboardShortcutsRegister() {
+		registerShortcut( {
+			name: 'core/block-editor/move-down',
+			category: 'block',
+			description: __( 'Move block Down.' ),
```
```suggestion
			description: __( 'Move the selected block(s) down.' ),
```

mcsf commented on docs/designers-developers/faq.md:
```diff
@@ -178,6 +178,16 @@ This is the canonical list of keyboard shortcuts:
 			<td><kbd>⌥</kbd><kbd>⌘</kbd><kbd>Y</kbd></td>
 		</tr>
+		<tr>
+			<td>Move the selected block(s) up.</td>
+			<td><kbd>Ctrl</kbd>+<kbd>Alt</kbd>+<kbd>Shift</kbd>+<kbd>T</kbd></td>
+			<td><kbd>⌥</kbd><kbd>⌘</kbd><kbd>⇧</kbd><kbd>T</kbd></td>
+		</tr>
+		<tr>
+			<td>Move the selected block(s) down.</td>
```
See below suggestions for revised labels.

ntsekouras committed:
Change shortcut description ( move up )

Co-authored-by: Miguel Fonseca <miguelcsf@gmail.com>

ntsekouras committed:
Change shortcut description ( move down )

Co-authored-by: Miguel Fonseca <miguelcsf@gmail.com>

ntsekouras committed:
Remove castArray from selected client ids

gziolo commented on packages/e2e-test-utils/src/index.js:
```diff
@@ -35,6 +35,10 @@ export { installPlugin } from './install-plugin';
 export { isInDefaultBlock } from './is-in-default-block';
 export { loginUser } from './login-user';
+export {
+	multiSelectBlocksByIds,
```
Could we consolidate those two methods into one API? Those methods are exposed as public API and added to the documentation. We will have to maintain them for long because the 3rd party project will use it.

ntsekouras replied:
I was talking with @mcsf about making the e2e tests more `user` interactive and not multi select `programmatically`, so I'll remove them as they will not be used. Do you think it has any value to keep them for now, if not used?

gziolo replied:
Not at all, we shouldn't keep methods that aren't used (unless they were part of public API previously).

gziolo commented on packages/e2e-test-utils/src/multi-select-blocks.js:
````diff
@@ -0,0 +1,65 @@
+ *
+ * @example
+ * ```js
+ * multiSelectBlocksByRange( 2, 3 );
+ * ```
+ * // selects three blocks -- from the second block to fourth (2, 3, 4).
````
```suggestion
 * multiSelectBlocksByRange( 2, 4 );
 * // selects blocks at index 2, 3, 4
```

How about we change the API to passing `startIndex` and `endIndex` to remove the need to use code comments to explain how it works?

gziolo commented on packages/e2e-test-utils/src/are-pre-publish-checks-enabled.js:
```diff
@@ -6,8 +6,8 @@ import { wpDataSelect } from './wp-data-select';
+ * @return {Promise<boolean>} Boolean which represents the state of prepublish checks.
  */
-export function arePrePublishChecksEnabled() {
+export async function arePrePublishChecksEnabled() {
```
Good catch 👍

gziolo commented on packages/e2e-test-utils/src/wp-data-select.js:
```diff
@@ -5,9 +5,9 @@
+ * @return {Promise<?Object>} Result of querying.
  */
-export function wpDataSelect( store, selector, ...parameters ) {
+export async function wpDataSelect( store, selector, ...parameters ) {
```
Great catch, `page.evaluate` that is used is async:

https://github.com/puppeteer/puppeteer/blob/main/docs/api.md#pageevaluatepagefunction-args

It might make some tests more stable. /cc @youknowriad

youknowriad replied:
I don't  think it makes any difference to add "async" here since page.evalute returns a promise already

ntsekouras replied:
It doesn't make any difference but it's explicit that returns a `Promise`. So someone would know they have to `await`.

mcsf replied:
In any case, the function signature was wrong, so this is a good catch.

ntsekouras committed:
Simplify e2e tests + remove multiSelect e2e util

mcsf commented on packages/e2e-tests/specs/editor/various/block-editor-keyboard-shortcuts.test.js:
```diff
@@ -0,0 +1,69 @@
+	describe( 'move blocks', () => {
+		const moveUp = async () => pressKeyWithModifier( 'secondary', 't' );
+		const moveDown = async () => pressKeyWithModifier( 'secondary', 'y' );
+		const multiSelectBlocks = async () => {
+			await page.keyboard.down( 'Shift' );
+			await page.keyboard.press( 'ArrowUp' );
+			await page.keyboard.up( 'Shift' );
+		};
```
`moveUp` and `moveDown` are unequivocal in their effects, and there is only one possible means to trigger each of them, so they can make sense as functions.

In contrast, `multiSelectBlocks` isn't unique: there are many ways to cause multiple blocks to be selected, and so this function depends on specific motions (in this case, upward). This means that when a particular test calls `multiSelectBlocks`, it's not clear for the reader what is happening in terms of editor state, unless they go up to read the definition of `multiSelectBlocks`.

This isn't a hard "no", but it's my own interpretation of what can improve the readability of tests.

ntsekouras replied:
You're not wrong about this. I changed it be more readable.

mcsf reviewed (commented):
The tests look much better, thanks for improving them. :)

ntsekouras committed:
minor e2e tests refactor

ntsekouras commented on packages/e2e-tests/specs/editor/various/block-editor-keyboard-shortcuts.test.js:
```diff
@@ -0,0 +1,69 @@
+			it( 'should move the block up', async () => {
+				await createTestParagraphBlocks();
+				expect( await getEditedPostContent() ).toMatchSnapshot();
+				await Promise.all( [ moveUp(), moveUp() ] ); // press twice
```
Thanks! Good note!

ntsekouras committed:
delete wpDataDispatch util

paaljoachim commented:
Here is an associated PR which just got merged: 
https://github.com/WordPress/gutenberg/pull/22453

mcsf reviewed (approved):
Nice work!

mcsf commented:
> Here is an associated PR which just got merged:
> **Prototype: move blocks between levels with keyboard**
> #22453

As the title suggests, there's still some stuff to iron out around the "Move to …" feature. The two PRs aren't mutually exclusive, so let's merge this one.

I see #22453 as more compelling as an accessibility improvement and as a QoL improvement when editing complex arrangements with nested blocks. Meanwhile, this PR falls more in the [writing flow](https://github.com/WordPress/gutenberg/labels/%5BFeature%5D%20Writing%20Flow) category, satisfying the need for quick motion around the selected block.

m0rg5 commented:
was this abandoned?

ntsekouras commented:
>was this abandoned?

@m0rg5 this has been merged for a while now..
